File: from-3101-to-3200/3152-special-array-ii/solution.py
class Solution:
    def isArraySpecial(self, nums: List[int], queries: List[List[int]]) -> List[bool]:
        result = [False] * len(queries)
        preSum = [0] * len(nums)
        preSum[0] = 1
        for i in range(1, len(nums)):
            preSum[i] = preSum[i - 1]
            if nums[i] % 2 != nums[i - 1] % 2:
                preSum[i] += 1

        for i, q in enumerate(queries):
            s, e = q
            if preSum[e] - preSum[s] == e - s:
                result[i] = True
        return result
        

File: from-3101-to-3200/3152-special-array-ii/solution.go
func isArraySpecial(nums []int, queries [][]int) []bool {
    result := make([]bool, len(queries))
    pairs := make([]int, len(nums) - 1)
    for i := 1; i < len(nums); i ++ {
        pairs[i - 1] = (nums[i - 1] % 2) ^ (nums[i] % 2)
    }
    preSum := make([]int, len(nums))
    preSum[0] = 1
    for i, val := range pairs {
        preSum[i+1] = preSum[i] + val
    }

    for i, q := range queries {
        if preSum[q[1]] - preSum[q[0]] == q[1] - q[0] {
            result[i] = true
        }
    }
    return result
}